### PR TITLE
table.fill: Bounds check before mutating the table

### PR DESCRIPTION
**Pull request by fitzgen**
Fixes https://github.com/WebAssembly/reference-types/issues/105

@rossberg please take a look. Thanks!

### CODE REVIEW

File: document/core/exec/instructions.rst [hunk 1]
```diff
@@ -651,27 +651,27 @@ Table Instructions
...
+17. Push the value :math:`\val` to the stack.
 
-    f. Push the value :math:`\I32.CONST~(n-1)` to the stack.
+18. Push the value :math:`\I32.CONST~(n-1)` to the stack.
 
-    c. Execute the instruction :math:`\TABLEFILL~x`.
+19. Execute the instruction :math:`\TABLEFILL~x`.
 
 .. math::
    \begin{array}{l}
```
Comment (binji) [inline document/core/exec/instructions.rst:677]: The formal notation needs to be fixed too -- this will still partially fill.

Comment (fitzgen) [reply document/core/exec/instructions.rst:677]: Perhaps someone else could fix that part? I'm not super familiar with latex / our conventions for the formal notation.

**Comment by rossberg** [comment]
Thanks for looking into this. The corresponding formal rules should look something like this:
```
.. math::
   \begin{array}{l}
   S; F; (\I32.\CONST~i)~\val~(\I32.\CONST~n)~(\TABLEFILL~x)
     \quad\stepto\quad S; F; \TRAP
     \\ \qquad
     \begin{array}[t]{@{}r@{~}l@{}}
     (\iff & i + n > |S.\STABLES[F.\AMODULE.\MITABLES[x]].\TIELEM|) \\
     \end{array}
   \\[1ex]
   S; F; (\I32.\CONST~i)~\val~(\I32.\CONST~0)~(\TABLEFILL~x)
     \quad\stepto\quad S; F; \epsilon
     \\ \qquad
     (\otherwise)
   \\[1ex]
   S; F; (\I32.\CONST~i)~\val~(\I32.\CONST~n+1)~(\TABLEFILL~x)
     \quad\stepto\quad S; F;
       \begin{array}[t]{@{}l@{}}
       (\I32.\CONST~i)~\val~(\TABLESET~x) \\
       (\I32.\CONST~i+1)~\val~(\I32.\CONST~n)~(\TABLEFILL~x) \\
       \end{array}
     \\ \qquad
     (\otherwise) \\
   \end{array}
```

**Comment by fitzgen** [comment]
> The corresponding formal rules should look something like this:

Thanks! I've added that to the commit.

**Comment by fitzgen** [comment]
(Note: I don't have merge permissions myself)

**Pull request merged**